packages/nuxt/src/app/composables/once.ts

```diff
@@ -0,0 +1,28 @@
+import { useNuxtApp } from '../nuxt'
+
+/**
+ * An SSR-friendly utility to call a method once
+ * @param key a unique key ensuring the function can be properly de-duplicated across requests
+ * @param fn a function to call
+ * @see https://nuxt.com/docs/api/utils/call-once
+ */
+export function callOnce (key?: string, fn?: (() => any | Promise<any>)): Promise<void>
+export function callOnce (fn?: (() => any | Promise<any>)): Promise<void>
+export async function callOnce (...args: any): Promise<void> {
+  const autoKey = typeof args[args.length - 1] === 'string' ? args.pop() : undefined
+  if (typeof args[0] !== 'string') { args.unshift(autoKey) }
+  const [_key, fn] = args as [string, (() => any | Promise<any>)]
+  if (!_key || typeof _key !== 'string') {
+    throw new TypeError('[nuxt] [callOnce] key must be a string: ' + _key)
+  }
+  if (fn !== undefined && typeof fn !== 'function') {
+    throw new Error('[nuxt] [callOnce] fn must be a function: ' + fn)
+  }
+  const nuxt = useNuxtApp()
+  // If key already ran
+  if (nuxt.payload.once.has(_key)) {
+    return
+  }
+  await fn()
+  nuxt.payload.once.add(_key)
```

Comment:
Because the key isn't added until after the function completes, it's possible for the function to be run more than once if called again before it finishes.

Instead, we should probably store the promise while it is running and return the original promise if `callOnce` is called again while running.

Reply:
@Atinux I've added an implementation - see what you think.

Reply:
Good for me! I thought about it too, thank you for implementing it ❤️